salt/modules/virt.py

```diff
@@ -2500,16 +2501,439 @@ def revert_snapshot(name, vm_snapshot=None, cleanup=False, **kwargs):
 
 def _capabilities(conn):
     '''
-    Return connection capabilities
-    It's a huge klutz to parse right,
-    so hide func for now and pass on the XML instead
+    Return connection capabilities as XML.
     '''
     caps = conn.getCapabilities()
     caps = minidom.parseString(caps)
 
     return caps
 
 
+def _get_xml_first_element_by_tag_name(node, name):
+    '''
+    Convenience function getting the first result of getElementsByTagName() or None.
+    '''
+    nodes = node.getElementsByTagName(name)
+    return nodes[0] if nodes else None
+
+
+def _get_xml_element_text(node):
+    '''
+    Get the text value of an XML element.
+    '''
+    return "".join([child.data for child in node.childNodes if child.nodeType == xml.dom.Node.TEXT_NODE])
+
+
+def _get_xml_child_text(node, name, default):
+    '''
+    Get the text value of the child named name of XML element node
+    '''
+    result = "".join([_get_xml_element_text(child) for child in node.childNodes
+                      if child.nodeType == xml.dom.Node.ELEMENT_NODE and child.tagName == name])
+    return result if result else default
+
+
+def _caps_add_machine(machines, node):
+    '''
+    Parse the <machine> element of the host capabilities and add it
+    to the machines list.
+    '''
+    maxcpus = node.getAttribute('maxCpus')
+    canonical = node.getAttribute('canonical')
+    name = _get_xml_element_text(node)
+
+    alternate_name = ""
+    if canonical:
+        alternate_name = name
+        name = canonical
+
+    machine = machines.get(name)
+    if machine is None:
```

Comment:
Can be `machine == ''` or `machine == 0`? If that is not the case, then `if not machine:` is the way to go here.